src/PluginsManager/Microsoft.Performance.Toolkit.PluginsManager.Core/Packaging/PluginPackage.cs

```diff
@@ -1,16 +1,305 @@
 ﻿// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+using System.IO.Compression;
+using System.IO;
+using System;
 using Microsoft.Performance.Toolkit.PluginsManager.Core.Packaging.Metadata;
+using Microsoft.Performance.SDK;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Threading;
 
 namespace Microsoft.Performance.Toolkit.PluginsManager.Core.Packaging
 {
     /// <summary>
     ///     Represents a read-only plugin package.
     /// </summary>
-    /// TODO: #236
-    public sealed class PluginPackage
+    public sealed class PluginPackage : IDisposable
     {
+        private readonly ZipArchive zip;
+        private static readonly string pluginContentPath = "plugin/";
+        private static readonly string pluginMetadataFileName = "pluginspec.json";
+        private bool disposedValue;
+        
+        /// <summary>
+        ///     Creates an instance of <see cref="PluginPackage"/>.
+        /// </summary>
+        /// <param name="fileName">
+        ///     The full file name of the plugin package.
+        /// </param>
+        public PluginPackage(string fileName)
+            : this(OpenFile(fileName))
+        {
+        }
+
+        /// <summary>
+        ///     Creates an instance of <see cref="PluginPackage"/>.
+        /// </summary>
+        /// <param name="stream">
+        ///     Stream for reading the plugin package file.
+        /// </param>
+        public PluginPackage(Stream stream) 
+            : this(stream, false)
+        {
+        }
+
+        /// <summary>
+        ///     Creates an instance of <see cref="PluginPackage"/>.
+        /// </summary>
+        /// <param name="stream">
+        ///     Stream for reading the plugin package file.
+        /// </param>
+        /// <param name="leaveOpen">
+        ///     <c>true</c> to leave <paramref name = "stream" /> open after <see cref="PluginPackage"/> is disposed.
+        ///     <c>false</c> otherwise.
+        /// </param>
+        public PluginPackage(Stream stream, bool leaveOpen)
+        {
+            Guard.NotNull(stream, nameof(stream));
+
+            try
+            {
+                this.zip = new ZipArchive(stream, ZipArchiveMode.Read, leaveOpen);
+                this.Entries = this.zip.Entries.Select(e => new PluginPackageEntry(e)).ToList().AsReadOnly();
+                this.PluginMetadata = ReadMetadata();
+
+            }
+            catch when (!leaveOpen)
```

Comment:
I would do
```cs
try 
{

}
finally
{
if (!leaveOpen)
  stream. Dispose();
}
```

Reply:
Actually, this should only run when there's an exception.  The stream should remain live while the ziparchive is live. `leaveOpen` is used to indicate whether the stream **after** the zipArchive is disposed.